homeassistant/components/switchbot/config_flow.py

```diff
@@ -92,28 +83,36 @@ async def async_step_user(
         errors = {}
 
         if user_input is not None:
-            user_input[CONF_MAC] = user_input[CONF_MAC].replace("-", ":").lower()
+            await self.async_set_unique_id(user_input[CONF_MAC].replace(":", ""))
+            self._abort_if_unique_id_configured()
 
-            # abort if already configured.
-            for item in self._async_current_entries():
-                if item.data.get(CONF_MAC) == user_input[CONF_MAC]:
-                    return self.async_abort(reason="already_configured_device")
+            user_input[CONF_SENSOR_TYPE] = SUPPORTED_MODEL_TYPES[
+                self._discovered_devices[self.unique_id]["modelName"]
+            ]
 
-            try:
-                return await self._validate_mac(user_input)
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
 
-            except NotConnectedError:
-                errors["base"] = "cannot_connect"
+        try:
+            self._discovered_devices = await self._get_switchbots()
 
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Unexpected exception")
-                return self.async_abort(reason="unknown")
+        except NotConnectedError:
+            errors["base"] = "cannot_connect"
+
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            return self.async_abort(reason="unknown")
+
+        supported_devices = {
+            device["mac_address"]
+            for device in self._discovered_devices.values()
+            if device.get("modelName") in SUPPORTED_MODEL_TYPES
```

Comment:
Exclude devices that are already setup and abort if no supported devices are discovered

Example

https://github.com/home-assistant/core/blob/dev/homeassistant/components/yeelight/config_flow.py#L166

Reply:
Thank you! Added code to compare the list with already configured devices and exclude them.

Reply:
Config flow also aborts if no devices match the condition.